### PR TITLE
CORE-8874: Add MGM client certificate subject to the saved group policy file

### PR DESCRIPTION
**Pull request by yift-r3**
API PR in https://github.com/corda/corda-api/pull/799

Adds the MGM client certificate subject to the group policy file - only in MTLS.  Parse the client certificate subject in the member group policy parser.

### CODE REVIEW

File: components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsServiceProcessor.kt [hunk 1]
```diff
@@ -244,16 +248,41 @@ class MemberOpsServiceProcessor(
             val tlsType = TlsType.fromString(
                 persistedGroupPolicyProperties.parseOrNull(PropertyKeys.TLS_TYPE, String::class.java)
             ) ?: TlsType.ONE_WAY
+            val mgmCertificateSubject = if (tlsType == TlsType.MUTUAL) {
+                val info = locallyHostedIdentitiesService.getIdentityInfo(holdingIdentity)
+                    ?: throw GroupPolicyGenerationException(
+                        "Mgm is not locally hosted. "+
+                        "If it had been configured, please retry the registration in a few seconds. "
```
Comment (dimosr) [inline components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsServiceProcessor.kt]: I think this message is not accurate here, it needs to refer to group policy generation.

Comment (dimosr) [reply components/membership/membership-service-impl/src/main/kotlin/net/corda/membership/service/impl/MemberOpsServiceProcessor.kt]: Also, one more thing: for the sake of keeping everything a bit more consistent, would it be better extracting the subject and persisting it as part of the group policy as part of the network setup call? And at this point, just read it from the group policy similar to the other properties?

**Review by dimosr — COMMENTED** [review]
Just one comment

File: components/membership/certificates-client-impl/src/main/kotlin/net/corda/membership/certificate/client/impl/MtlsMgmClientCertificateKeeper.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import java.security.cert.X509Certificate
+
+internal class MtlsMgmClientCertificateKeeper(
+    private val membershipGroupReaderProvider: MembershipGroupReaderProvider,
+    private val membershipPersistenceClient : MembershipPersistenceClient,
+    private val membershipQueryClient : MembershipQueryClient,
+    private val layeredPropertyMapFactory: LayeredPropertyMapFactory,
+    private val certificateFactory: CertificateFactory = CertificateFactory.getInstance("X.509"),
+) {
+    fun keepSubjectIfNeeded(
```
Comment (dimosr) [inline components/membership/certificates-client-impl/src/main/kotlin/net/corda/membership/certificate/client/impl/MtlsMgmClientCertificateKeeper.kt]: I had a hard time understanding what the method does from the name without looking at the code.. Isn't really the goal of this to add the subject when needed (e.g. mutual TLS / MGM etc.)? 

How about something like `addMGMCertSubjectToGroupPolicy()` for method name?

File: components/membership/certificates-client-impl/src/test/kotlin/net/corda/membership/certificate/client/impl/MtlsMgmClientCertificateKeeperTest.kt [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            keeper.keepSubjectIfNeeded(
+                mgmHoldingIdentity,
+                groupProperty,
+                "certificate",
+            )
+        }
+    }
+
+    @Test
+    fun `keepSubjectIfNeeded will throw an the certificate is invalid`() {
```
Comment (dimosr) [inline components/membership/certificates-client-impl/src/test/kotlin/net/corda/membership/certificate/client/impl/MtlsMgmClientCertificateKeeperTest.kt]:
```suggestion
    fun `keepSubjectIfNeeded will throw an exception if the certificate is invalid`() {
```

**Review by dimosr — DISMISSED** [review]

**Review by charlieR3 — DISMISSED** [review]

**Review by charlieR3 — APPROVED** [review]

**Review by mcgovc — APPROVED** [review]
LGTM - build

**Pull request merged**